### PR TITLE
time boost in folds generation

### PR DESCRIPTION
**Pull request by aldder**
image

### CODE REVIEW

File: tscv/_split.py [hunk 1]
```diff
@@ -161,14 +161,27 @@ def __complement_masks(self, masks):
...
+
     def __complement_indices(self, indices, n_samples):
         before, after = self.gap_before, self.gap_after
         for index in indices:
-            complement = np.arange(n_samples)
-            for i in index:
-                begin = max(i - before, 0)
-                end = min(i + after + 1, n_samples)
-                complement = np.setdiff1d(complement, np.arange(begin, end))
+            complement = np.setdiff1d(
```
Comment (WenjieZ) [inline tscv/_split.py:175]: `index` might not be a contiguous set.

**Review by WenjieZ — CHANGES_REQUESTED** [review]
Please consider the corner case and resubmit.

File: tscv/_split.py (hunk 1 above)
Comment (aldder) [reply tscv/_split.py:175]: Hi, sorry, can you provide an example where index is not contiguous?

Comment (WenjieZ) [reply tscv/_split.py:175]: For example, the test set can be `[5, 7, 9]` or `[4, 6, 8]`, which are not contiguous. This kind of test sets is uncommon, but the interface does not forbid the user to do so. That's why we need to handle this corner case (which is actually a more general case).

Comment (aldder) [reply tscv/_split.py:175]: I'm sorry, still can't look a situation like this. in your example if we have a set like `[5, 7, 9]`, then we get the `min - before `and the `max + gap_after etc`, but then, in order to get the complement, we create a range (contiguous) between begin and end, so....

Comment (WenjieZ) [reply tscv/_split.py:175]: All right. Let's consider the following case.
```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9]
test = [1, 2, 8, 9]
gap_before = 1
gap_after = 1
```

In this case, the training set should be `[4, 5, 6]`, not empty.

Comment (WenjieZ) [reply tscv/_split.py:175]: All right. Let's consider the following case.
```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9]
test = [1, 2, 8, 9]
gap_before = 1
gap_after = 1
```

In this case, the training set should be `[4, 5, 6]`, not empty.

Comment (aldder) [reply tscv/_split.py:175]: Ok I got it, you're right, this case is not correctly handled with my code, but is more a general case, as you said before. I'm not sure that these kind of splits can be generate by the actual cv methods.

Comment (WenjieZ) [reply tscv/_split.py:175]: What do you mean by this sentence?
> I'm not sure that these kind of splits can be generate by the actual cv methods.

If you tweak your modification a little bit, you can handle the more general case without sacrificing too much the efficiency.

Comment (aldder) [reply tscv/_split.py:175]: I mean that, according to the implementation of `_iter_test_indices` methods, there shouldn't be cases of test folds with uncontiguous elements, but I may be wrong.
Btw I don't think I can find a tweak asap, and I don't have much time right now. Feel free to leave the PR open or delete if as you wish :)

Comment (WenjieZ) [reply tscv/_split.py:175]: > according to the implementation of `_iter_test_indices` methods, there shouldn't be cases of test folds with uncontiguous elements, but I may be wrong.

Indeed, all existing subclasses in this package give contiguous test sets. However, the base class `GapCrossValidator` does not impose this constraint. Users can subclass it and produce uncontiguous test sets:
```python
class Foo(tscv.GapCrossValidator):
    def _iter_test_indices(self, X=None, y=None, groups=None):
        n = len(X)
        yield list(range(0, n, 2))
        yield list(range(1, n, 2))
        
    def get_n_splits(self, X=None, y=None, groups=None):
        return 2

for train_index, test_index in Foo().split(range(10)):
    print("TRAIN:", train_index, "TEST:", test_index)
```

```
TRAIN: [1 3 5 7 9] TEST: [0, 2, 4, 6, 8]
TRAIN: [0 2 4 6 8] TEST: [1, 3, 5, 7, 9]
```

Comment (WenjieZ) [reply tscv/_split.py:175]: If you don't have time, I would suggest making a small improvement by using branching. If the test set is contiguous, then use your code. Otherwise, use the original code. You can test the contiguity with `len(index) == index[-1] - index[0] + 1`.

**Comment by aldder** [comment]
Hi @WenjieZ, I edited this PR, following some benchmarks:

**With contiguous test sets:**
```
cv_orig = GapKFold(n_splits=5, gap_before=1, gap_after=1)

for train_index, test_index in cv_orig.split(np.arange(10)):
    print("TRAIN:", train_index, "TEST:", test_index)

... TRAIN: [3 4 5 6 7 8 9] TEST: [0 1]
... TRAIN: [0 5 6 7 8 9] TEST: [2 3]
... TRAIN: [0 1 2 7 8 9] TEST: [4 5]
... TRAIN: [0 1 2 3 4 9] TEST: [6 7]
... TRAIN: [0 1 2 3 4 5 6] TEST: [8 9]
```
```
cv_opt = GapKFold(n_splits=5, gap_before=1, gap_after=1)

for train_index, test_index in cv_opt.split(np.arange(10)):
    print("TRAIN:", train_index, "TEST:", test_index)

... TRAIN: [3 4 5 6 7 8 9] TEST: [0 1]
... TRAIN: [0 5 6 7 8 9] TEST: [2 3]
... TRAIN: [0 1 2 7 8 9] TEST: [4 5]
... TRAIN: [0 1 2 3 4 9] TEST: [6 7]
... TRAIN: [0 1 2 3 4 5 6] TEST: [8 9]
```
```
%%timeit
folds = list(cv_orig.split(np.arange(10000)))
... 1.21 s ± 37.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
```
%%timeit
folds = list(cv_opt.split(np.arange(10000)))
... 4.74 ms ± 44.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**With uncontiguous test sets:**
```
cv_orig = _XXX_(_xxx_, gap_before=1, gap_after=1)

for train_index, test_index in cv_orig.split(np.arange(10)):
    print("TRAIN:", train_index, "TEST:", test_index)

... TRAIN: [5 6 7 8 9] TEST: [0 1 2 3]
... TRAIN: [7 8 9] TEST: [0 1 4 5]
... TRAIN: [3 4 9] TEST: [0 1 6 7]
... TRAIN: [3 4 5 6] TEST: [0 1 8 9]
... TRAIN: [0 7 8 9] TEST: [2 3 4 5]
... TRAIN: [0 9] TEST: [2 3 6 7]
... TRAIN: [0 5 6] TEST: [2 3 8 9]
... TRAIN: [0 1 2 9] TEST: [4 5 6 7]
... TRAIN: [0 1 2] TEST: [4 5 8 9]
... TRAIN: [0 1 2 3 4] TEST: [6 7 8 9]
```
```
cv_opt = _XXX_(_xxx_, gap_before=1, gap_after=1)

for train_index, test_index in cv_opt.split(np.arange(10)):
    print("TRAIN:", train_index, "TEST:", test_index)

... TRAIN: [5 6 7 8 9] TEST: [0 1 2 3]
... TRAIN: [7 8 9] TEST: [0 1 4 5]
... TRAIN: [3 4 9] TEST: [0 1 6 7]
... TRAIN: [3 4 5 6] TEST: [0 1 8 9]
... TRAIN: [0 7 8 9] TEST: [2 3 4 5]
... TRAIN: [0 9] TEST: [2 3 6 7]
... TRAIN: [0 5 6] TEST: [2 3 8 9]
... TRAIN: [0 1 2 9] TEST: [4 5 6 7]
... TRAIN: [0 1 2] TEST: [4 5 8 9]
... TRAIN: [0 1 2 3 4] TEST: [6 7 8 9]
```
```
%%timeit
folds = list(cv_orig.split(np.arange(10000)))
... 1.23 s ± 75.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
```
%%timeit
folds = list(cv_opt.split(np.arange(10000)))
... 4.78 ms ± 49.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
